opm/autodiff/MultisegmentWell.hpp

```diff
@@ -0,0 +1,336 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef OPM_MULTISEGMENTWELL_HEADER_INCLUDED
+#define OPM_MULTISEGMENTWELL_HEADER_INCLUDED
+
+
+#include <opm/autodiff/WellInterface.hpp>
+
+namespace Opm
+{
+
+    template<typename TypeTag>
+    class MultisegmentWell: public WellInterface<TypeTag>
+    {
+    public:
+        typedef WellInterface<TypeTag> Base;
+
+        // TODO: the WellState does not have any information related to segments
+        using typename Base::WellState;
+        using typename Base::Simulator;
+        using typename Base::IntensiveQuantities;
+        using typename Base::FluidSystem;
+        using typename Base::ModelParameters;
+        using typename Base::MaterialLaw;
+        using typename Base::BlackoilIndices;
+
+        /// the number of reservior equations
+        using Base::numEq;
+
+        // TODO: for now, not considering the polymer, solvent and so on to simplify the development process.
+        // TODO: should I begin with the old primary variable or the new fraction based variable systems?
+        // Let us begin with the new one
+        // TODO: we need to have order for the primary variables and also the order for the well equations.
+        // sometimes, they are similar, while sometimes, they can have very different forms.
+
+        // TODO: the following system looks not rather flexible. Looking into all kinds of possibilities
+        // TODO: gas is always there? how about oil water case?
+        // Is it gas oil two phase case?
+        static const bool gasoil = numEq == 2 && (BlackoilIndices::compositionSwitchIdx >= 0);
+        static const int GTotal = 0;
+        static const int WFrac = gasoil? -1000: 1;
+        static const int GFrac = gasoil? 1 : 2;
+        static const int SPres = gasoil? 2 : 3;
+
+        ///  the number of well equations // TODO: it should have a more general strategy for it
+        static const int numWellEq = GET_PROP_VALUE(TypeTag, EnablePolymer)? numEq : numEq + 1;
+
+        using typename Base::Scalar;
+        using typename Base::ConvergenceReport;
+
+        /// the matrix and vector types for the reservoir
+        using typename Base::Mat;
+        using typename Base::BVector;
+        using typename Base::Eval;
+
+        // sparsity pattern for the matrices
+        // [A C^T    [x       =  [ res
+        //  B  D ]   x_well]      res_well]
+
+        // the vector type for the res_well and x_well
+        typedef Dune::FieldVector<Scalar, numWellEq> VectorBlockWellType;
+        typedef Dune::BlockVector<VectorBlockWellType> BVectorWell;
+
+        // the matrix type for the diagonal matrix D
+        typedef Dune::FieldMatrix<Scalar, numWellEq, numWellEq > DiagMatrixBlockWellType;
```

Comment:
I should know this by heart now, but I often forget, FieldMatrix<double,3,3>::invert is notoriously unstable in version 2.5.0 and below. Therefore we might want to change this to
```
#if  DUNE_VERSION_NEWER(DUNE_ISTL, 2 , 5, 1)
        // 3x3 matrix block inversion was unstable from at least 2.3 until and
        // including 2.5.0
        // the matrix type for the diagonal matrix D
        typedef Dune::FieldMatrix<Scalar, numWellEq, numWellEq > DiagMatrixBlockWellType;
#else
        // the matrix type for the diagonal matrix D
        typedef Dune::MatrixBlock<Scalar, numWellEq, numWellEq > DiagMatrixBlockWellType;
#endif
```
There is also a specialized version of invert for 4x4 MatrixBlock in ISTLSolver.hpp which has travelled upstream. I asume that is used for the solvent/polymer case? @totto82, is it OK to copy that one to dune-common? But even then we only have it from version 2.6 or 2.5.2. Therefore the `#if` clause might still be to optimistic.

Maybe this change would solve the issue with the iterative solver compared to UMFPack? What matrix type is used there and where is it defined?

BTW a PR to fix this for StandardWells is in the making...

Reply:
Needs to be `DUNE_VERSION_NEWER_REV` instead of `DUNE_VERSION_NEWER`

Reply:
Thanks for the reply.  I will try `MatrixBlock` to see if it helps. 
FYI, For multisegment well without solvent or any other component, the block size for diagonal matrix is 4X4. The size of the diagonal matrix will depends on the number of the segment.  For example, if there is 10 segment, the size of the matrix will be 10 X 10, and block size will be 4X4. The matrix is non-symmetric.

Reply:
> BTW a PR to fix this for StandardWells is in the making...

I think you meant `StandardWell`. Sorry for the confusing naming, while `StandardWells` is going to disappear.

Reply:
> What matrix type is used there and where is it defined?

The matrix type is the `DiagMatrixBlockWellType` above.